pkg/controller/handlers/poweruserworkspace/poweruserworkspace.go

```diff
@@ -74,78 +72,143 @@ func (h *Handler) isPrivilegedRole(role types.Role) bool {
+		return h.reconcileWorkspaceRole(ctx, client, &workspaces.Items[0], effectiveRole)
+	}
+
+	// Case 4: Shouldn't have workspace and doesn't - no action needed
+	return nil
+}
+
+// createWorkspaceWithRole creates a PowerUserWorkspace with the given role.
+func (h *Handler) createWorkspaceWithRole(ctx context.Context, client kclient.Client, namespace string, user gatewaytypes.User, role types.Role) error {
+	userIDStr := strconv.Itoa(int(user.ID))
+
 	workspace := &v1.PowerUserWorkspace{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: namespace,
 			Name:      system.GetPowerUserWorkspaceID(userIDStr),
 		},
 		Spec: v1.PowerUserWorkspaceSpec{
 			UserID: userIDStr,
-			Role:   user.Role,
+			Role:   role,
 		},
 	}
 
 	return create.OrGet(ctx, client, workspace)
 }
 
-func (h *Handler) deleteWorkspaceForUser(ctx context.Context, client kclient.Client, namespace string, userID string) error {
-	var workspaces v1.PowerUserWorkspaceList
-	if err := client.List(ctx, &workspaces, &kclient.ListOptions{
+// deleteAllWorkspaces deletes all given workspaces.
+func (h *Handler) deleteAllWorkspaces(ctx context.Context, client kclient.Client, workspaces []v1.PowerUserWorkspace) error {
+	for _, workspace := range workspaces {
+		if err := client.Delete(ctx, &workspace); err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// reconcileWorkspaceRole ensures the workspace's role matches the effective role.
+func (h *Handler) reconcileWorkspaceRole(ctx context.Context, client kclient.Client, workspace *v1.PowerUserWorkspace, effectiveRole types.Role) error {
+	// If role already matches, nothing to do
+	if workspace.Spec.Role == effectiveRole {
+		return nil
+	}
+
+	oldRole := workspace.Spec.Role
+	workspace.Spec.Role = effectiveRole
+
+	if err := client.Update(ctx, workspace); err != nil {
+		return err
+	}
+
+	// If demoting from PowerUserPlus/Admin to PowerUser, clean up resources
+	if !effectiveRole.HasRole(types.RolePowerUserPlus) && oldRole.HasRole(types.RolePowerUserPlus) {
+		return h.cleanupWorkspaceResources(ctx, client, workspace)
+	}
+
+	return nil
+}
+
+// cleanupWorkspaceResources removes ACRs and MCPServers when demoting to PowerUser.
+func (h *Handler) cleanupWorkspaceResources(ctx context.Context, client kclient.Client, workspace *v1.PowerUserWorkspace) error {
+	namespace := workspace.Namespace
+
+	// Delete AccessControlRules in this workspace
+	var acrs v1.AccessControlRuleList
+	if err := client.List(ctx, &acrs, &kclient.ListOptions{
 		Namespace: namespace,
 		FieldSelector: fields.SelectorFromSet(map[string]string{
-			"spec.userID": userID,
+			"spec.powerUserWorkspaceID": workspace.Name,
 		}),
 	}); err != nil {
 		return err
 	}
 
-	for _, workspace := range workspaces.Items {
-		if err := client.Delete(ctx, &workspace); err != nil && !errors.IsNotFound(err) {
+	for _, acr := range acrs.Items {
+		if err := client.Delete(ctx, &acr); err != nil && !errors.IsNotFound(err) {
 			return err
 		}
 	}
```

Comment:
I think you should be able to make use of `client.DeleteAllOf` here instead of issuing individual delete calls.

Reply:
Fixed